### ISSUE TITLE
Module side effects

### ISSUE THREAD

**Issue by samhh** [jonaskello/eslint-plugin-functional#155]
This one probably can't be done, but I'm curious if anyone has any thoughts - can we lint against module side effects?

Only way I've thought of so far is to lint specifically against `fp-ts` and enforce that you can't call `Task` or `IO` functions in a module except within another function.

**Comment by RebeccaStevens** [comment]
I believe if you are using the rule `no-expression-statement`, this will prevent module side effects unless the side effect happens as a result of a function call.
If it becomes possible to [ensure a function is pure in typescript](https://github.com/microsoft/TypeScript/issues/7770) then this could also be prevented.

**Comment by jonaskello** [comment]
Yes, having a pure modifier for functions in typescript would be awesome :-). 

I usually try to avoid having any code outside of functions in all modules. Having side-effects happen just by importing a module is IMO never wanted. Perhaps we could check if there is anything else than function definitions in a module? That check would be less specific of course but it would eliminate any side-effect that happen on import. Or would that be too restrictive?
